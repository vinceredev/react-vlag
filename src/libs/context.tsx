import React, { createContext } from 'react';
import { useCookies } from 'react-cookie';

interface VlagPropsType {
  id: string;
  name: string;
  description?: string;
  helperText?: string;
  disabled: boolean;
}

type VlagProps = {
  children: React.ReactNode;
  flags: Array<VlagPropsType>;
};

interface VlagContextType {
  isEnabled: (id: string) => boolean;
  setEnable: (id: string, val: boolean) => void;
  getFlags: () => Array<VlagPropsType>;
  flagMap: Array<string>;
}

const VlagContext = createContext<VlagContextType>({
  isEnabled: () => false,
  setEnable: () => undefined,
  getFlags: () => [],
  flagMap: []
});

export const VlagProvider = ({ children, flags }: VlagProps) => {
  const flagMap = flags.map((fl) => fl.id);
  const [cookies, setCookie] = useCookies(flagMap);

  const isEnabled = (id: string) => {
    return cookies[id] === 'true';
  };

  const setEnable = (id: string, val: boolean) => {
    setCookie(id, val);
  };

  const getFlags = () => {
    return flags;
  };

  let hasDuplicate = flagMap.some((val, i) => flagMap.indexOf(val) !== i);
  if (hasDuplicate) {
    return <>flag id is duplicate</>;
  }

  return (
    <VlagContext.Provider value={{ isEnabled, setEnable, getFlags, flagMap }}>
      {children}
    </VlagContext.Provider>
  );
};

export const useVlag = () => React.useContext(VlagContext);
