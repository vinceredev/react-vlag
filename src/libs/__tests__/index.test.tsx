import React from 'react';
import renderer from 'react-test-renderer';
import { VlagProvider } from '../context';

it('success: vlag provider renders correctly', () => {
  const vlag = (
    <VlagProvider
      flags={[
        {
          id: 'new_checkout_page',
          name: 'Checkout Page',
          description:
            'Lorem Ipsum is simply dummy text of the printing and typesetting industry. ',
          helperText: 'Will be available for beta testing end of 2022',
          disabled: false
        },
        {
          id: 'new_product_page',
          name: 'Product Page',
          description:
            'Lorem Ipsum is simply dummy text of the printing and typesetting industry. ',
          disabled: false
        }
      ]}
    >
      content
    </VlagProvider>
  );

  const tree = renderer.create(vlag);
  expect(tree).toMatchSnapshot();
});

it('error: vlag provider renders fails if flags id same', () => {
  const vlag = (
    <VlagProvider
      flags={[
        {
          id: 'new_checkout_page',
          name: 'Checkout Page',
          description:
            'Lorem Ipsum is simply dummy text of the printing and typesetting industry. ',
          helperText: 'Will be available for beta testing end of 2022',
          disabled: false
        },
        {
          id: 'new_checkout_page',
          name: 'Checkout Page',
          description:
            'Lorem Ipsum is simply dummy text of the printing and typesetting industry. ',
          disabled: false
        }
      ]}
    >
      content
    </VlagProvider>
  );

  const tree = renderer.create(vlag);
  expect(tree).toMatchSnapshot();
});
