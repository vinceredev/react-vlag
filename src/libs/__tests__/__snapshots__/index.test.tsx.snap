// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error: vlag provider renders fails if flags id same 1`] = `"flag id is duplicate"`;

exports[`success: vlag provider renders correctly 1`] = `"content"`;
